### ISSUE TITLE
v2 routing ambiguity in 'Nested URLs without Layout Nesting'

### ISSUE THREAD

**Issue by calldanfeng** [remix-run/remix#5846]
### What version of Remix are you using?

1.14.3

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

Given the routes in following structure: 
```
app/routes
└── boxes_
    ├── bottles_
    │   ├── drinks_
    │   │   └── route.tsx
    │   └── route.tsx
    ├── candies_
    │   └── route.tsx
    └── route.tsx
```
To make the test simple, each route.tsx exports a default function which returns the name of its business job.  e.g. 'Boxes', 'Bottles', 'Drinks', 'Candies'.

And visit 
1. `http://localhost:3000/boxes/candies` 
2. `http://localhost:3000/boxes/bottles/drinks`

---
Remarks:
In my app, nested layout is rarely used, but nested URLs are used frequently and they may also be defined deeply.   According to the doc [Nested URLs without Layout Nesting](https://remix.run/docs/en/1.14.3/file-conventions/route-files-v2#md-nested-urls-without-layout-nesting), developers can achieve a non-nested layout with a trailing underscore.  But actually it also requires its parent route NOT having a trailing underscore. 
In order to make it work, I have to either define the route structure like:
```
app/routes
├── boxes
│   └── route.tsx
└── boxes_
    ├── bottles
    │   └── route.tsx
    ├── bottles_
    │   └── drinks
    │       └── route.tsx
    └── candies
        └── route.tsx

```
or like:
```
app/routes
└── boxes
    ├── _index
    │   └── route.tsx
    ├── bottles
    │   ├── _index
    │   │   └── route.tsx
    │   └── drinks
    │       └── _index
    │           └── route.tsx
    └── candies
        └── _index
            └── route.tsx
```
Personally feeling, they are less self-contained or clear than the origin one. 

### Expected Behavior

Visit 1. should return 'Candies'
Visit 2. should return 'Drinks'.

### Actual Behavior

Visit 1. returns 'Boxes'
Visit 2. returns 'Boxes'.

**Comment by MajorTom327** [comment]
I'm not sure you're currently using v2 convention actually.

Have you put the flag https://remix.run/docs/en/1.14.3/file-conventions/route-files-v2#md-route-file-naming-v2 on ?

Next, you should have this 
```
app/routes
├── boxes_.bottles_.drinks_/index.tsx
├── boxes_.bottles_/index.tsx
├── candies_/index.tsx
└── _index.tsx
```

| file name | route |
| --- | --- |
| `boxes_.bottles_.drinks_/index.tsx` | `/boxes/bottles/drinks` |
| `boxes_.bottles_/index.tsx` | `/boxes/bottles` |
| `candies_/index.tsx` | `/candies` |
| `_index.tsx` | `/` |


Hope that help ! :)

**Comment by calldanfeng** [comment]
> I'm not sure you're currently using v2 convention actually.
> 
> Have you put the flag https://remix.run/docs/en/1.14.3/file-conventions/route-files-v2#md-route-file-naming-v2 on ?
> 
> Next, you should have this
> 
> ```
> app/routes
> ├── boxes_.bottles_.drinks_/index.tsx
> ├── boxes_.bottles_/index.tsx
> ├── candies_/index.tsx
> └── _index.tsx
> ```
> 
> file name	route
> `boxes_.bottles_.drinks_/index.tsx`	`/boxes/bottles/drinks`
> `boxes_.bottles_/index.tsx`	`/boxes/bottles`
> `candies_/index.tsx`	`/candies`
> `_index.tsx`	`/`
> Hope that help ! :)

Thanks @MajorTom327 for the quick response!  I have turned v2 routing on.  

I was testing if nested folders are supported in v2 routing.  I was worrying about the plain structure of v2 routing when there are lots of routers defined and the efforts for refactoring a level of routes. 
 
Let's make the question simple, and say the way you define the routes are a 'happy path' test case for a developer defining v2 routings.  Are the 'sad paths'  I listed above supported in v2 routing?

**Comment by logikaljay** [comment]
As a remix newbie, I didn't understand the v2 folder based routing, I thought it worked as @calldanfeng described.

Partly because the docs were not overly clear, partly because what I was doing was working.

I had:

```
routes/
routes/settings
routes/settings/route.tsx // <- nested layout for the navigation on the "settings" segment
routes/settings/project
routes/settings/project/route.tsx // <- project settings page
routes/settings/roles
routes/settings/roles/route.tsx // <- roles settings page
```

It wasn't until I added `routes/settings/project/create/route.tsx` that I noticed it was full of the content from the `routes/settings/project/route.tsx` page.

A weird google search later, I found this thread and saw @MajorTom327's image above. Restructured my routes folder and all is working now as expected.

**Comment by kiliman** [comment]
So the new v2 flat-routes convention has two modes:

## flat-files
With flat-files, the *entire* route name is the filename. You cannot use folders, even for organization. The way Remix determines nesting is via common shared prefix. If a parent route doesn't include an `<Outlet>` then you denote that with a trailing `_`.

### Examples:
routes/settings.project.tsx
routes/settings.project.create.tsx  <- parent layout is settings.project (longest common prefix)
routes/settings.project_.create.tsx <- we don't want project as parent (add trailing _)

## flat-folders
Once you add a folder anywhere in your route name, you now opt-in to flat-folders. Remix doesn't care about the actual folders. It treats it as if the `/` was `.` like flat-files. The main difference is that the folder path is the route name. Remix only cares about the final filename. The route file is route.tsx (or index.tsx). Any other file is ignored. This is useful for colocating files with your route, like components, images, CSS, etc.

### Examples
routes/settings/projects/route.tsx
routes/settings/project.create/route.tsx  <- mix of folder and flat files
routes/settings/project.create/component.tsx  <- ignored, not route or index

**Comment by calldanfeng** [comment]
@kiliman explains the v2 routing behavior clearly. 

But I found it mentally taxing to define the folders if I don't have any nested layout requirement in my application.  Say I add the '_' tail in every folders just hope to get rid of nested layouts. However, the following folder structure will not work as expected.
```
routes/settings_/projects_/route.tsx
routes/settings_/projects_/create_/route.tsx
```
/settings/projects/create will still take the layout of the parent path /settings/projects. To fix this, remove the '_' tailing in the parent path. 
```
routes/settings_/projects/route.tsx
routes/settings_/projects_/create_/route.tsx
```
I'm not sure it's by design or a bug of v2 routing...

**Comment by brophdawg11** [comment]
This seems to be functioning as designed, and I think is being a bit obfuscated by the usage of the `route.tsx` convention.

Without that, you'd use the following to get no layout nesting:

| Route File | URL |
|--|--|
| `routes/settings.tsx` | `/settings` |
| `routes/settings_/projects.tsx` | `/settings/projects` | 
| `routes/settings_/projects_/create.tsx` | `/settings/projects/create` |

The trailing slash is therefore used when an _existing matching route file exists_ and you do not want to nest inside of that.

@machour Did you have a specific bug fix in mind for v2 here?

**Comment by cairin** [comment]
@brophdawg11 I mentioned something similar in #5930 .

I think it boils down to this being documented incorrectly/poorly. If you look at the preparing for v2 docs, there is a comparison showing how the various things change from v1 to v2.

For the **Nested URLs without Layout Nesting** line, [the v2 version](https://github.com/remix-run/remix/blob/main/docs/pages/v2.md?plain=1#L104) is actually incorrect and doesn't translate to how it works.

```txt bad
routes
└── dashboard.calendar.$projectId.print.tsx
```
In v1 this would produce `/dashboard/calendar/some-projecti-d/print` with absolutely no layout nesting. But to achieve that in v2 you would need the following file structure:

```diff
routes
-└── dashboard_.calendar.$projectId.print.tsx
+└── dashboard_.calendar_.$projectId_.print.tsx
```
This is because in the current implementation of v2, if any other route exists it will try nest in that route - and in that specific example would try nest in `├── dashboard.calendar.tsx`. Thus making the dot notation explicit opt-in to layout nesting, where as in v1 the dot notation was explicit opt-out of layout nesting.

**Comment by cairin** [comment]
And here is a reproduction of this in action:

https://stackblitz.com/edit/remix-run-remix-932gqa?file=app/routes/dashboard_.nested.deeply.again.tsx

As you can see if you route to `/dashboard/nested/deeply/again` it should render **You'll never see this** but instead it renders **This'll render fine :)** because it's trying to nest inside `dashboard_.nested.deeply.tsx`, which doesn't have an `<Outlet />` so it just shows it's own page. If you were to put an `<Outlet />` in `dashboard_.nested.deeply.tsx` then `dashboard_.nested.deeply.again.tsx` would render in that outlet.

**Comment by kiliman** [comment]
One of the main differences between v1 and v2 routing is that before, dotted filenames could be taken in isolation, meaning there was no nesting for this path. You'd place the route file in a folder if you wanted to nest.

However, with v2 routing, folders are not used for nesting, so the filename itself has to encode that parent/child structure. That means the filename is dependent on the _surrounding context_. So `a.b.c.tsx` can perfectly define a route without any nested layouts. That is until you add another route, `a.b.tsx` or even `a.tsx`. Then your existing route must be renamed if they want to maintain the same non-nested layout.

It's basically the same as creating folders and moving files around. Once you understand the rules, it shouldn't be an issue. But I do feel that this should be called out in the documentation.

**Comment by cairin** [comment]
The only material difference betweent the two is if you look at the app directory you now see `routes/a.b.c.tsx` instead of `routes/a/b/c.tsx` - to achieve the same functionality. 

Why is this so confusing then? Well, simply put, it's just **not obvious anymore**. The filename was always dependent on the _surrounding context_, it was just obvious which routes previous developers intended to be nested. Because they had put them in a directory. Without any context you could look at a project and immediately know which routes had an `<Outlet />` in them.

This has a few impacts on developer experience:
- Developers now need to assume all route segments are nested, always.
- When a developer looks at their directory structure, there is no way to tell which segments are intended to be nestable routes (The only way to know this now is to look in each file for `<Outlet />` or `useOutlet()`). Before, our IDE's would handle this for us because they can natively delineate files in directories thanks to the file system. 
- Every time a developer creates a route, they now need to check **all** potential file matches for nested routes. Because it's not obvious which routes might accomodate nesting and which might not.
- Adding routes isn't just additive. Developers now have to worry about breaking code they aren't touching.
- If a developer wants an escape hatch out of nesting entirely, they need to put trailing underscores on _all_ path segments. And to ensure they don't have to think about this again in the future they have to always do this. On every file.

Sure, documenting this could help, and people will learn this set of rules and make it work. But why make it harder in the first place?

The same developer experience as v1 could easily be achieved with flat routes if the **nesting was visually opt-in** as before.

**Comment by kiliman** [comment]
You can continue to use the v1 routing by installing the new compat package:

https://github.com/remix-run/v1-compat-utils/tree/main/packages/v1-route-convention

I also have the package `remix-flat-routes`, which is much more flexible with its routing than v2.

**Comment by brophdawg11** [comment]
@cairin I think the typo in the docs is actually the following which might explain the confusion.  There is no `dashboard/calendar/$projectId` UI so I think the intent was to differentiate the "edit" view of a project from the "print" view.  I'll get this fixed up shortly 👍 

```diff
-└── dashboard.calendar.$projectId.print.tsx
+└── dashboard.projects.$projectId.print.tsx
```

```diff
-└── dashboard_.calendar.$projectId.print.tsx
+└── dashboard_.projects.$projectId.print.tsx
```

With that, you do not need to escape _each_ section of the route, just escaping the root layout is enough:

<img width="807" alt="Screenshot 2023-05-15 at 3 11 28 PM" src="https://github.com/remix-run/remix/assets/1609022/08825e3f-2162-4a86-b989-48fc7d66285b">
<br/><br/>

Beyond that I think this thread is veering into subjective territory:
 
> Well, simply put, it's just not obvious anymore

This has not been a common bit of feedback thus far.  Some folks (as expected) prefer the nested folder approach which is why we made sure to keep that usable via `@remix-run/v1-route-convention`.  Remix provides the `routes` option in `remix.config.js` for exactly this reason as well - we've found that there is simply no convention that pleases everyone, so we picked one we think is fairly simple and works in _most_ cases.  But devs are always free to create their own conventions and even publish them as @kiliman has done with his excellent `remix-flat-routes` package.

**Comment by brophdawg11** [comment]
I updated the "Preparing for V2" docs in #6397.  We're always open to additional PRs to the docs that might help clarify cases like this as well!

**Comment by duongductrong** [comment]
> 

That means we will always create a new route in routes (flatten routes), right guy?

For example, I want to create authentication ver2, it looks like:

routes/_auth.login.tsx <--- auth old ver1
routes/_auth.sign-up.tsx <--- auth old ver1
routes/_auth.tsx <--- auth old layout ver1
routes/_auth.login-v2.tsx <--- new 
routes/_auth.sign-up.tsx <--- new

Is that true? they always nested on routes and no nested more than 2 level folders, right?

> Well, in v2, as I understand it, they will prefer this that way. That doesn't mean the v1 will be disabled. It should be available. But the official way will be by not nesting.
> 
> I'm currently working on a big app with many sub-routes. And personnaly, at first I was sceptic. But now, it feel pretty good. As I can found my ressources easily and they are grouped in vs code by topic (As sub-routes start by the same.)
> 
> Plus, when you refactor your routes. Renaming a folder or file is easier than moving structure of folder.
> 
> Take a look about this route folder with not that many folders but a pretty good number for a PoC:
> 
> <img alt="image" width="345" src="https://user-images.githubusercontent.com/16199396/226363169-f66d0ad5-c8be-49ba-93f3-ecda6095fe7c.png">
> 
> (It's a basic accounting app)
> 
> I don't think having sub-routes inside of v2 should be a great thing. That will mean we will have a "v1.5" I think you should be one way or one other, but not both. In a long run having both will be a mess and inconsistent.

**Comment by brophdawg11** [comment]
This was released in [1.11.0](https://github.com/remix-run/remix/releases/tag/remix%401.11.00) via [this proposal](https://github.com/remix-run/remix/discussions/4482).  You can continue using the old pattern via the npm package if you prefer that convention, or if you have questions you can ask them in the discord: https://rmx.as/discord.
